### ISSUE TITLE
DISTINCT and DISTINCTON generate wrong and complex queries

### ISSUE THREAD

**Issue by nlhkh** [slick/slick#1340]
Suppose I have a simple query like this using `hello-slick-3.1.0`

``` scala
val query = for {
    (c, s) <- coffees joinLeft suppliers distinctOn(_._1.name)
} yield(c, s)

println(query.result.statements)
db.run(query result)
```

One would naturally think that the equivalent sql query is

``` sql
SELECT DISTINCT ON (c.name) *
FROM coffees c
LEFT JOIN suppliers s
ON c.sup_ID = s.id
```

This query is simple and execute fast. However, the generated query by Slick is as follow

``` sql
select distinct x2."STATE", x2."SUP_NAME", x2."CITY", x2."STREET", x2."ZIP", x3."PRICE", x3."SALES", x3."COF_NAME", x2."SUP_ID", x3."TOTAL", x3."SUP_ID", (case when (x2."SUP_ID" is null) then null else 1 end) from "COFFEES" x3 left outer join "SUPPLIERS" x2 on 1=1
```

This query execute very slow because it misses the optimization of the database with this syntax. In my case when the tables have 10000 rows each, the query timeout after several minutes. On some rare cases the slow query completes in 3 minutes, as opposed to about 200ms with the hand written query.

I think this is a bug since all it has to do is to create a DISTINCT ON statement with the given projection.

**Comment by lisinge** [comment]
I'm having the same problem. This would really help making slow queries fast as they should be

**Comment by nlhkh** [comment]
For now you can use plain SQL, which is not too bad actually.

**Comment by pocman** [comment]
I'm having an issue with `distinctOn` using play-slick 1.1.1 and postgresql 9.4.1207
I'm trying to do this : http://stackoverflow.com/a/19028149/4768481
I wrote : 
```
for {
      assignment: Tables.Assignments <- Assignments.sortBy(assignment => (assignment.orderid, assignment.createdat.desc)).distinctOn(_.orderid)
    } yield assignment
```

The generated sql statement is : 
`select distinct on ("contractid", "userid", "orderid", "workingscheduleid", "id", "createdat") "orderid" as x4, "createdat" as x8, "contractid" as x6, "userid" as x7, "workingscheduleid" as x5, "id" as x3 from "assignments" order by "orderid", "createdat" desc;`

**Comment by pocman** [comment]
As @nlhkh suggested, I used plain SQL.
```    
val allLatestIdsQuery: DBIO[Seq[String]] = sql"""select distinct on ("orderid") "id" from "assignments" order by "orderid", "createdat" desc""".as[String]
```

**Comment by pocman** [comment]
Hi,
still have this issue, interestingly, the order is not the same for the projections.

`Assignments.distinctOn(a => a.orderUuid).sorted(assignment => (assignment.orderUuid, assignment.createdAt.desc)).map(_.uuid).result`
generates 
`select distinct on ("working_schedule_uuid", "order_uuid", "uuid", "created_at", "contract_uuid", "user_id") "uuid" from "assignments" order by "order_uuid", "created_at" desc;`

`Assignments.distinctOn(a => a).sorted(assignment => (assignment.orderUuid, assignment.createdAt.desc)).map(_.uuid).result`
generates 
`select distinct on ("uuid", "order_uuid", "working_schedule_uuid", "contract_uuid", "user_id", "created_at") "uuid" from "assignments" order by "order_uuid", "created_at" desc;`

**Comment by dnagoda** [comment]
Just wanted to add a curious behavior that I'm seeing with MySQL and distinctOn. Given a query with a distinctOn before several joins and filters I noticed inconsistent ordering of the arguments in the resulting inner GROUP BY clause resulting in a later Join on the wrong column.

```
SELECT 
    COUNT(1)
FROM
    (SELECT 
        `DISTRIBUTOR_NUMBER` AS x2,
            `DISTRIBUTOR_USER_NUMBER` AS x3,
            MIN(`SALES_REP_NUMBER`) AS x4,
            MIN(`SALES_REP_NAME`) AS x5
    FROM
        `sales_dump`
    WHERE
        `CTL_DUMP_ID` = 2
    GROUP BY `DISTRIBUTOR_NUMBER` , `DISTRIBUTOR_USER_NUMBER`) x6
        INNER JOIN
    `company` x7 ON x7.`distributor_number` = x6.x2
        LEFT OUTER JOIN
    `user_company_v2` x8 ON (x8.`company_id` = x7.`company_id`)
        AND (x8.`distributor_user_number` = x6.x3)
WHERE
    x8.`user_company_id` IS NULL
```

on a subsequent call

```
SELECT 
    COUNT(1)
FROM
    (SELECT 
        `DISTRIBUTOR_USER_NUMBER` AS x2,  -- **THIS HAS BEEN REORDERED**
            `DISTRIBUTOR_NUMBER` AS x3,
            MIN(`SALES_REP_NUMBER`) AS x4,
            MIN(`SALES_REP_NAME`) AS x5
    FROM
        `sales_dump`
    WHERE
        `CTL_DUMP_ID` = 2
    GROUP BY `DISTRIBUTOR_NUMBER` , `DISTRIBUTOR_USER_NUMBER`) x6
        INNER JOIN
    `company` x7 ON x7.`distributor_number` = x6.x2  -- **BUT THIS JOIN IS THE SAME**
        LEFT OUTER JOIN
    `user_company_v2` x8 ON (x8.`company_id` = x7.`company_id`)
        AND (x8.`distributor_user_number` = x6.x3)
WHERE
    x8.`user_company_id` IS NULL
```

I'll try to create a test case.

**Comment by szeiger** [comment]
We don't have any optimization for lifting a DISTINCT out of a JOIN. With https://github.com/slick/slick/pull/1649 it should be possible to do the lifting manually in order to get a properly optimized query.

**Comment by szeiger** [comment]
Correction: I got confused by operator precedence (I prefer to use method notation). The `distinct` call is already outside of the join, so this should work as is with https://github.com/slick/slick/pull/1649.

**Comment by nafg** [comment]
I'm stilling seeing this in slick 3.2.0

**Comment by lisinge** [comment]
We are also seeing this on slick 3.2.0

**Comment by lisinge** [comment]
Any updates on this? Being able to use `distinctOn` would be really helpful

**Comment by missingfaktor** [comment]
For those who might stumble upon this in future, here is a workaround we are using: https://gist.github.com/missingfaktor/aa6c264c5b7411fa48a6a5b654dd0917.

**Comment by coding4m** [comment]
Also seeing this on slick 3.2.0, any update on this?

**Comment by lukaszwierzynski** [comment]
Confirm, query:

`val query = locations
        .distinctOn(_.trailerId)
        .sortBy(_.createdAt desc)
        .filter(_.trailerId.inSet(ids))
        .result`

ends with:
`select x2.x3, x2.x4, x2.x5, x2.x6, x2.x7, x2.x8, x2.x9, x2.x10, x2.x11, x2.x12, x2.x13 from (select distinct on ("updated_at", "zip", "street_number", "city", "country", "latitude", "longitude", "created_at", "id", "trailer_id", "street") "updated_at" as x13, "zip" as x8, "street_number" as x7, "city" as x5, "country" as x9, "latitude" as x11, "longitude" as x10, "created_at" as x12, "id" as x3, "trailer_id" as x4, "street" as x6 from "locations" order by "created_at" desc) x2 where x2.x4 in ('10215873-ab98-45fc-a16b-f26945aaa8bc', 'bdac6026-617f-4288-9dd4-96c3b87a312c')`

**Comment by zrss** [comment]
cc

**Comment by wim82** [comment]
Also being affected by this.. I spent quite some time only to realise the generated query is wrong.

**Comment by ewiner** [comment]
Can we at least relabel this to a `bug` instead of being a `improvement` in the "Feature ideas" milestone?  I think it's clearly a bug to have a `.distinctOn()` function that apparently discards its arguments and doesn't end up as `DISTINCT ON` in SQL (or an equivalent statement).

**Comment by shmishleniy** [comment]
If someone stuck with this, here is possible solution, not merged yet
https://github.com/slick/slick/pull/2097

**Comment by renatocaval** [comment]
Fixed in #2097 and released. Available on v3.3.3
